src/main/flight/autopilot.c

```diff
@@ -91,14 +158,223 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+        // separate PID controllers for latitude (NorthSouth or ns) and longitude (EastWest or ew)
+
+        // ** P **
+
+        const float nsP = nsDistance * positionPidCoeffs.Kp;
+        const float ewP = ewDistance * positionPidCoeffs.Kp;
+
+        // ** I **
+
+        const float leak = 1.0f - 0.4f * posHold.gpsDataIntervalS; // gpsDataIntervalS is not more than 1.0s
+        if (!posHold.isStartingNS){
+            // only accumulate iTerm after completing the start phase
+            // perhaps need a timeout on the start phase ?
+            posHold.integralNS += nsDistance * posHold.gpsDataIntervalS;
+        } else {
+            // while moving sticks, slowly leak iTerm away, approx 2s time constant
+            posHold.integralNS *= leak;
+        }
+        if (!posHold.isStartingEW){
+            posHold.integralEW += ewDistance * posHold.gpsDataIntervalS;
+        } else {
+            posHold.integralEW *= leak;
+        }
+
+        const float nsI = posHold.integralNS * positionPidCoeffs.Ki;
+        const float ewI = posHold.integralEW * positionPidCoeffs.Ki;
+
+        // ** D ** //
+        // get change in distance in NS and EW directions from gps.c using the `GPS_distances` function
+        // this gives cleaner velocity data than the module supplied GPS Speed and Heading information
+
+        const float velocityNS = (nsDistance - posHold.previousDistanceNS) * posHold.gpsDataFreqHz; // cm/s, minimum step 11.1 cm/s
+        const float velocityEW = (ewDistance - posHold.previousDistanceEW) * posHold.gpsDataFreqHz;
+        posHold.previousDistanceNS = nsDistance;
+        posHold.previousDistanceEW = ewDistance;
+
+        float accelerationNS = (velocityNS - posHold.previousVelocityNS) * posHold.gpsDataFreqHz;
+        posHold.previousVelocityNS = velocityNS;
+        float accelerationEW = (velocityEW - posHold.previousVelocityEW) * posHold.gpsDataFreqHz;
+        posHold.previousVelocityEW = velocityEW;
+
+        // scale and filter - filter cutoffs vary during the startup phase
+        float nsD = velocityNS * positionPidCoeffs.Kd;
+        pt1FilterUpdateCutoff(&velocityNSLpf, posHold.pt1Gain);
+        nsD = pt1FilterApply(&velocityNSLpf, nsD);
+
+        float ewD = velocityEW * positionPidCoeffs.Kd;
+        pt1FilterUpdateCutoff(&velocityEWLpf, posHold.pt1Gain);
+        ewD = pt1FilterApply(&velocityEWLpf, ewD);
+
+        float nsA = accelerationNS * positionPidCoeffs.Kf;
+        pt2FilterUpdateCutoff(&accelerationNSLpf, posHold.pt1Gain); // using PT1 gain for stronger cutoff
+        nsA = pt2FilterApply(&accelerationNSLpf, nsA);
+
+        float ewA = accelerationEW * positionPidCoeffs.Kf;
+        pt2FilterUpdateCutoff(&accelerationEWLpf, posHold.pt1Gain); // using PT1 gain for stronger cutoff
+        ewA = pt2FilterApply(&accelerationEWLpf, ewA);
+
+        // limit sum of D and A because otherwise can be too aggressive when starting at speed
+        float nsDA = nsD + nsA;
```

Comment:
In effect, you are implementing PIDD regulator, but naming acceleration part `Kf`.  It is quite confusing.

Reply:
I did not want to use the kF factor for the Acceleration term.  @KarateBrot had suggested re-using `pidCoefficient_t`, in which the last entry is `kF`. To have `kA`, we would need to define a new table I guess?  

Yes the controller has D and an 'acceleration' element that is basically just a 'quicker' D element, to some extent overcoming the filtering delay.  And since the filter does some integration, the A element is not so dissimilar to D.

Adding the acceleration element is definitely very useful, since lag is our great enemy here.  Whether the acceleration factor needs a separate user-adjustable PID gain is something I'm not sure about.  It is in the CLI now mostly for testing.

At this point it appears that assigning the same gain to acceleration as we have for D works very well.  In that case we can remove the CLI setting.

Also now the lag per-axis doesn't matter like before, we can probably add them earlier, to simplify the code.  I'll look at that soon.

Reply:
Common way to do this is to use PD controller for speed (that can be tuned first) and PI controller over position, with speed as output. 
It may have some other advantages - position controller can run at slower rate (and avoid upsampling filter), speed controlled may use ACC data to get faster feedback etc ..

ad lag: for LTI system, it does not matter in which order filters are applied (and if filtering value before or after calculation of difference)